lib/Graph/TensorLayout.cpp

```diff
@@ -0,0 +1,557 @@
+      return layout->getSerializedLayout();
+    }
+  }
+  if (auto *RN = llvm::dyn_cast<ReshapeNode>(node)) {
+    if (auto *layout = getLayoutFromName(RN->getLayout(),
+                                         layoutNameToLayoutDescription_)) {
+      return layout->getSerializedLayout();
+    }
+    auto result = node->getNthResult(n);
+    auto *user = (*result.getUsers().begin()).getUser();
+    int inputIdx = getInputIdx(user, result);
+    if (inputIdx >= user->getNumInputs() || llvm::isa<TransposeNode>(user)) {
+      return getLayoutsForDims()[dims.size()].getSerializedLayout();
+    }
+    auto layout = getNthInputLayoutRequirements(user, inputIdx);
+    if (auto *layoutDesc =
+            getLayoutFromName(layout, layoutNameToLayoutDescription_)) {
+      return layoutDesc->getSerializedLayout();
+    }
+  }
+  return getLayoutsForDims()[dims.size()].getSerializedLayout();
+}
+
+bool TensorLayoutCommon::isSatisfiedBy(
+    TypeRef ty, const TensorLayoutDescription &destLayout,
+    const TensorLayoutDescription *srcLayout) const {
+  // Strides of the type (in elements).
+  auto strides = ty->strides();
+  if (strides.size() != destLayout.getNumDims()) {
+    return false;
+  }
+  unsigned idx = 0;
+  for (const auto &dim : destLayout.getDims()) {
+    // dim.alignment is in bytes, but strides are in elements.
+    if (strides[idx] * ty->getElementSize() % destLayout.getAlignment(dim) !=
+        0) {
+      return false;
+    }
+    idx++;
+  }
+  if (!srcLayout) {
+    return true;
+  }
+  if (destLayout.getNumDims() != srcLayout->getNumDims()) {
+    return false;
+  }
+  // Names should be compatible. * is compatible to anything.
+  if (srcLayout->getSerializedLayout().size() !=
+      destLayout.getSerializedLayout().size()) {
+    return false;
+  }
+  for (unsigned idx = 0, e = destLayout.getSerializedLayout().size(); idx < e;
+       ++idx) {
+    // '*' is compatible with anything.
+    if (destLayout.getSerializedLayout()[idx] == '*' ||
+        srcLayout->getSerializedLayout()[idx] == '*') {
+      continue;
+    }
+    // Non-'*' are only compatible with themselves.
+    if (srcLayout->getSerializedLayout()[idx] ==
+        destLayout.getSerializedLayout()[idx]) {
+      continue;
+    }
+    return false;
+  }
+  return true;
+}
+
+static std::string returnBaseReqOrNHWC(std::string baseReq) {
+  auto baseReqHelper = TensorLayoutDescription(baseReq);
+  if (!baseReqHelper.isSameLayout(
+          CanonicalTensorLayout::getInstance().getLayoutsForDims()[4])) {
+    return baseReq;
+  }
+  // NHWC is the canonical default
+  return CanonicalTensorLayout::getInstance().getDefaultNDLayout(4);
+}
+
+std::string
+CanonicalTensorLayout::getNthInputLayoutRequirements(const Node *node,
+                                                     size_t n) {
+  auto baseReq = TensorLayoutCommon::getNthInputLayoutRequirements(node, n);
+  if (acceptsAnyLayout(node)) {
```

Comment:
Should acceptsAnyLayout take n for the inputs/outputs number? Or do we say that all such nodes are always layout agnostic on all of their inputs/outputs?

Reply:
I refactored it: it accepts *any* for all input and output just for data parallel nodes, however that is not correct for nodes that can accept multiple layouts as input and propagate it to their output in canonical, but not all, backends. I did not add an `n` but I did make that clearer and split it into two functions.